### PR TITLE
Improve SessionCodeChecks to detect better the ALREADY_LOGGED_IN situation

### LINKED ISSUE #19677: AlreadyLoggedIn when impersonating a user in a SAML client
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

core

### Describe the bug

Checking the issue #15337 I noticed that in the current main branch impersonation does not work when using a SAML client. I think that the admin user should be in the same realm than the impersonated user too. After some debugging the problem is that the AUTH_SESSION_ID is not returned for the new impersonated session and this creates issues.

The bug is intoduced in 20.0.2 and I think that probably triggered by this commit https://github.com/keycloak/keycloak/pull/15981/commits/a1ab33778df468d5b9c1828918eedfe7faca1722. 

I'm going to send a little PR that creates the root authentication session and assigns the AUTH_SESSION_ID cookie when doing an impersonation. This way everything seems to to work.

### Version

21.0.2

### Expected behavior

The SAML application should work with the impersonation.

### Actual behavior

Message `You are already logged in` is displayed when going to keycloak login page to authenticate.

### How to Reproduce?

Install a SAML client (in my case is a EAP 7.4 with SAML adapters installed). And do the following:

* Login in the realm console with an admin user that can impersonate. The admin should be in the same realm (master in my case).
* Go to users an impersonate another user.
* The accuount console works.
* Go to the SAML application.
* It redirects to the login page and the error `You are already logged in` is displayed.

As commented I think the reason is commit https://github.com/keycloak/keycloak/commit/6086201fe0472b89eb3d4e188bd4132c29f8d4c9, which checked AUTH_SESSION_ID cookie.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by rmartinc**
Closes https://github.com/keycloak/keycloak/issues/19677

In some situations the message `You are already logged in` appears if an incorrect AUTH_SESSION_COOKIE is present when re-login (in #19677 the situation is impersonation). I'm changing the order of that check in the `SessionCodeChecks` class. Now the already logged in situation is only checked if the restart cookie response is not found (a login or re-login is not in progress). Impersonate endpoint also expires AUTH_SESSION_ID cookie just in case. A new test is added that emulates the SAML situation after impersonation.

All tests pass in my env.  Let's see the CI.

### CODE REVIEW

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java [hunk 1]
```diff
@@ -181,35 +179,26 @@ public AuthenticationSessionModel initialVerifyAuthSession() {
...
-        }
+            if (authResult != null && authResult.getSession() != null) {
+                LoginFormsProvider loginForm = session.getProvider(LoginFormsProvider.class).setAuthenticationSession(authSession)
+                        .setSuccess(Messages.ALREADY_LOGGED_IN);
 
-        if (userSession != null) {
-            LoginFormsProvider loginForm = session.getProvider(LoginFormsProvider.class).setAuthenticationSession(authSession)
-                    .setSuccess(Messages.ALREADY_LOGGED_IN);
+                if (client == null) {
+                    loginForm.setAttribute(Constants.SKIP_LINK, true);
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:195]: I guess this is not strictly related but a nice improvement to how the page is rendered, right?

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java [hunk 2]
```diff
@@ -181,35 +179,26 @@ public AuthenticationSessionModel initialVerifyAuthSession() {
...
-
-        boolean authenticating = !CookieHelper.getCookieValue(session, AUTH_SESSION_ID).isEmpty();
+        // Otherwise just try to restart from the cookie
+        RootAuthenticationSessionModel existingRootAuthSession = authSessionManager.getCurrentRootAuthenticationSession(realm);
+        response = restartAuthenticationSessionFromCookie(existingRootAuthSession);
 
-        if (authenticating) {
-            // if there is an auth session, make sure the user is not yet authenticated
+        // if restart from cookie was not found check if the user is already authenticated
+        if (response.getStatus() != Response.Status.FOUND.getStatusCode()) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:187]: It looks more reliable than just checking the presence of auth cookie. However, I'll wait for @mposolda feedback to see if there is some corner case that we might be missing.

It is not so nice to make the decision based on status code but I don't have a better idea other than refactoring more code and making the fix a bit more complex.

**Review by pedroigor — DISMISSED** [review]
@rmartinc Thanks for incorporating the changes. Let's see what @mposolda has to say.

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java (hunk 1 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:195]: @pedroigor I have just moved this part from above, so I'm Doing the same than before. :smile:

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java (hunk 2 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:187]: I think it's OK now. if the response is not found it means there is no restart cookie, so there is no login (or re-login) in progress. Only in that case the check for identity cookie and already authenticated is performed.

With the previous order, the check for identity cookie was done if there was an AUTH_SESSION_ID cookie, and that created the problem for the impersonation. If you check the test, it's not even an impersonation, the AUTH_SESSION_ID cookies are removed, that triggered the problem at the second re-login because a new AUTH_SESSION_ID was generated but the auth session was immediately removed. In general the issue is triggered if the AUTH_SESSION_ID cookie is invalid (wrong, invalid, expired or whatever).

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java (hunk 1 above)
Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:195]: Right, missed it.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java [hunk 3]
```diff
@@ -1773,6 +1779,39 @@ public void testAdapterRoleMappings() throws Exception {
...
+        assertSuccessfullyLoggedIn(salesPostSigServletPage, "principal=bburke");
+        // remove session cookies for the application to force re-login against the keycloak server
+        driver.manage().deleteAllCookies();
+
+        // go to the login-actions/authenticate page to remove auth session cookies
+        authenticate.setAuthRealm(SAMLSERVLETDEMO);
+        authenticate.navigateTo();
+        waitForPageToLoad();
+        assertCurrentUrlStartsWith(authenticate);
+        assertThat(driver.getPageSource(), containsString("You are already logged in."));
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java:1794]: Minor: It's a bit better to use something like:
```
infoPage.assertCurrent();
Assert.assertEquals("You are already logged in.", infoPage.getInfo());
```
I think it is safer option regarding various browsers etc.

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java (hunk 2 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:187]: +1 to add some "Impersonation integration" test instead of current `testReloginWithAndWithoutAuthSessionCookie` . The small issue with the current test is, that it uses lots of "implementation details" of Keycloak (for example manually opening auth pages, manually removing cookies etc), which typical user won't do. However at the same time, I understand that sometimes it is quite hard to test "real integration" .

My preference would be also to have "real impersonation" test, but if you confirm that it is impossible/unreliable to test, I am fine with current automated test as a fallback.

**Review by mposolda — COMMENTED** [review]
@rmartinc @pedroigor The fix itself looks good to me, Thanks for it. I've added minor comment to the test, but in general, I second @pedroigor comment about "real impersonation" test. Do you think it is possible?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java (hunk 3 above)
Comment (rmartinc) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java]: Done!

File: services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java (hunk 2 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:187]: I have done something, but it's not perfect. We cannot use selenium directly because we need the admin console to really use impersonation and that needs a real web driver (chrome, firefox,...). The default one is not enough. I have followed the spirit in the `ImpersonationTest` and done a impersonation using an HttpClient, getting all the cookies, set the cookies in the selenium driver and test the SAML app. I did it in the SAML tests because a SAML client is needed. Let's see the tests but it works OK to me.

Comment (rmartinc) [reply services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java:187]: One more thing. I have maintained the other test (modified a bit) to test the general situation when the AUTH_SESSION_ID cookie is invalid.

**Review by pedroigor — APPROVED** [review]
